### PR TITLE
Reset spyder and restart from within running application

### LINKED ISSUE #2401: Unable to reset settings from the Main Window
Hi,

I have tried reseting the User interface i.e. windows layout from view but I cannot get back to original. 
I have removed the Spyder-Py2.app and reinstalled it but it doesn't restore to default. Also, I have tried using Anaconda but still the same problem. 
screen shot 2015-05-04 at 6 39 11 pm

Any suggestion on proper removal of Spyder or setting it to default. Thanks.

### PR DESCRIPTION
**Pull request by goanpeca**
Fixes #2401

---
## Description

Adds a new entry on the tools menu for reseting spyder from within the application

image

Or as an option in preferences... decision pending
image
After clicking a message box will ask to confirm, probably the message can be more explanatory so the user knows what he\she is about to do.

image

## Todo

- [x] Agree on location of option
- [x] Agree on text of the new option

### CODE REVIEW

**Comment by Nodd** [comment]
I would put it in the general tab in the preferences window.

**Comment by goanpeca** [comment]
That feels weird, cause the preferences tab defines behaviors not actions.

This is an action, and it would be nice if it has a straightforward place for it.

**Comment by Nodd** [comment]
Yeah, but it is an action applied to the behaviors !
It could be added as another pushbutton along with Apply, OK and Cancel.

**Comment by goanpeca** [comment]
In that case I prefer adding the extra button.

---

Opinions @spyder-ide/core-developers?

**Comment by goanpeca** [comment]
@blink1073, @SylvainCorlay, @ccordoba12 any comments on this one?

**Comment by blink1073** [comment]
I like it as part of the Tools menu, as it is a global action.

**Comment by goanpeca** [comment]
@sylvaincorlay? @ccordoba12 comments?

**Comment by Nodd** [comment]
I really think that resetting the preferences should be done from the preference window. That's the same reason why resetting shortcuts is done from the shortcuts tab.

**Comment by goanpeca** [comment]
I understand my dear, that is why I said, if it was going to be there, then it should be a button next to apply, ok, cancel, etc....

But I would like to hear from @SylvainCorlay and @ccordoba12

**Comment by goanpeca** [comment]
@Nodd there might be a reason to keep the restart option outside. Sometimes if there is an error or conflict the preferences dialog might not open, in which case restarting and reseting could not be done from within the gui.

**Comment by Nodd** [comment]
#2412 should lower the probability of this happening. Also it's still possible to use the commandline switch or remove the directory by hand  (of course it's less user-friendly).

**Comment by goanpeca** [comment]
Ok @spyder-ide/core-developers this is the other option for the button.

This is a very small PR I would like to get off my back so please take a look at:

- Location. Tools menu or Preferences dialog?, or both.
    - image

    - image

- Text of message
- Any comments on code?

Cheers

File: spyderlib/plugins/configdialog.py [hunk 1]
```diff
@@ -943,4 +950,4 @@ def apply_settings(self, options):
         if self.main.historylog is not None:
             self.main.historylog.apply_plugin_settings(['color_scheme_name'])
         if self.main.inspector is not None:
-            self.main.inspector.apply_plugin_settings(['color_scheme_name'])
+            self.main.inspector.apply_plugin_settings(['color_scheme_name'])
```
Comment (Nodd) [inline spyderlib/plugins/configdialog.py]: Missing a newline here

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: oops

Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: Fixed

**Comment by goanpeca** [comment]
@Nodd Fixed

File: spyderlib/restart_app.py [hunk 2]
```diff
@@ -135,6 +144,29 @@ def main():
...
+        command_reset = '"{0}" "{1}" {2}'.format(python, spyder, args_reset)
+
+        # Try to reset
+        try:
+            p = subprocess.Popen(command_reset, shell=shell)
+            p.communicate()
+            pid_reset = p.pid
+        except Exception as error:
+            print(command)
+            print(error)
```
Comment (Nodd) [inline spyderlib/restart_app.py]: Using the `traceback` module would be nicer here, since it prints the full traceback as if the Exception wasn't catched. You can use [`traceback.print_exc()`](https://docs.python.org/3.4/library/traceback.html#traceback.print_exc) with no arguments, it prints the right thing at the right place.

File: spyderlib/restart_app.py [hunk 3]
```diff
@@ -135,6 +144,29 @@ def main():
...
+            time.sleep(15)
```
Comment (Nodd) [inline spyderlib/restart_app.py]: 15s, really ?

File: spyderlib/restart_app.py [hunk 4]
```diff
@@ -135,6 +144,29 @@ def main():
...
+
+        # Wait for reset process to end before restarting
+        while True:
```
Comment (Nodd) [inline spyderlib/restart_app.py]: This is potentially an infinite loop. Should we limit it to 30s or something, and warn the suer that there was an error ?
On the other hand the possibility of infinite loop is really small...

File: spyderlib/spyder.py [hunk 5]
```diff
@@ -713,6 +713,9 @@ def create_edit_action(text, tr_text, icon):
                                                   module_completion.reset(),
                                         tip=_("Refresh list of module names "
                                               "available in PYTHONPATH"))
+            reset_spyder_action = create_action(self,
+                                        _("Reset Spyder to factory defaults"),
```
Comment (Nodd) [inline spyderlib/spyder.py]: That's not a good pep8 alignment here

File: spyderlib/restart_app.py (hunk 3 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: The idea was to be able to see the command line if an error occurs... otherwise it will close too fast... I could reduce to 10... ?

File: spyderlib/restart_app.py (hunk 2 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Great, I will do :)

File: spyderlib/spyder.py (hunk 5 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: Ok, will fiix :)

File: spyderlib/restart_app.py (hunk 4 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Well actually it can happen, good catch, so maybe hmmmm 50? that would be like 10 seconds... but it might take longer to restart on windows ... sometimes windows is weird...and takes a long time for a (spyder) process to die

**Comment by goanpeca** [comment]
@Nodd so you suggest something regarding the "infinite" loop?... or should we just let it roll and see how it works?

File: spyderlib/spyder.py [hunk 6]
```diff
@@ -713,6 +713,9 @@ def create_edit_action(text, tr_text, icon):
                                                   module_completion.reset(),
                                         tip=_("Refresh list of module names "
                                               "available in PYTHONPATH"))
+            reset_spyder_action = create_action(
+                self, _("Reset Spyder to factory defaults"),
+                triggered=lambda: self.reset_spyder())
```
Comment (Nodd) [inline spyderlib/spyder.py]: You can avoid the lambda here I think, using `self.reset_spyder` directly

File: spyderlib/plugins/configdialog.py [hunk 7]
```diff
@@ -124,13 +124,17 @@ class ConfigDialog(QDialog):
     
     def __init__(self, parent=None):
         QDialog.__init__(self, parent)
-        
+
+        self._parent = parent
```
Comment (Nodd) [inline spyderlib/plugins/configdialog.py]: I think you don't need this line, since `self.main` is used.

**Comment by Nodd** [comment]
I think the correct behaviour is to wait something like 30s, and if it's not finished, kill the process, stop the restart and display a QMessageBox with an error.
I'm afraid that if we let the infinite loop it'll bite us sometimes in the future for whatever reason...

File: spyderlib/restart_app.py [hunk 8]
```diff
@@ -135,6 +145,29 @@ def main():
...
+
+        # Try to reset
+        try:
+            p = subprocess.Popen(command_reset, shell=shell)
+            p.communicate()
+            pid_reset = p.pid
+        except Exception as error:
+            print(command)
+            traceback.print_exc()
+            time.sleep(15)
```
Comment (Nodd) [inline spyderlib/restart_app.py]: If there is a problem here, the restart should be stopped instead of waiting 15s. This way the error will show up in the internal console.
Actually that means that you don't need to use `traceback.print_exc()`, just reraise the exception!

File: spyderlib/spyder.py (hunk 6 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: True!

File: spyderlib/plugins/configdialog.py (hunk 7 above)
Comment (goanpeca) [inline spyderlib/plugins/configdialog.py]: Gotcha

File: spyderlib/restart_app.py (hunk 8 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: I am not sure I get you... by internal console you mean? at this point we already left the initial spyder instance and we are inside the restart script. I will add an extra flag so that the restart stops and a QMessageBox displays something.

**Comment by goanpeca** [comment]
Ok @Nodd I added a series of checks for better error handling which include a display of a message box in case errors occur, including close, reset and restart errors. Errors catched in the try,except part will also display the actual error. Errors based on too long running time will not display the error, but a friendly message....

File: spyderlib/restart_app.py [hunk 9]
```diff
@@ -129,18 +180,60 @@ def main():
     shell = os.name != 'nt'
 
     # Wait for original process to end before launching the new instance
-    while True:
+    max_counter = 1500  # Max loops to perform before aborting
+    counter = 0
+    while counter < max_counter:  # This is equivalent to ~ 5 minute (1500*0.2)
```
Comment (Nodd) [inline spyderlib/restart_app.py]: A `for` loop maybe ? ;)

File: spyderlib/restart_app.py [hunk 10]
```diff
@@ -129,18 +180,60 @@ def main():
...
-    while True:
+    max_counter = 1500  # Max loops to perform before aborting
+    counter = 0
+    while counter < max_counter:  # This is equivalent to ~ 5 minute (1500*0.2)
         if not is_pid_running(pid):
             break
         time.sleep(0.2)  # Throttling control
+        counter += 1
+
+    if counter == max_counter:
```
Comment (Nodd) [inline spyderlib/restart_app.py]: With a for loop you can just use `else` here. The `else` part is run if the loop is not broken.

Comment (goanpeca) [inline spyderlib/restart_app.py]: You read my mind... I was just doing this

File: spyderlib/restart_app.py [hunk 11]
```diff
@@ -129,18 +180,60 @@ def main():
...
+            p.communicate()
+            pid_reset = p.pid
+        except Exception as error:
+            p.kill()
+            error_message(type_='reset', error=error)
+
+        counter = 0
+        max_counter = 50
+        # Wait for reset process to end before restarting
+        while counter < max_counter:  # This is ~= 10 seconds (50*0.2)
```
Comment (Nodd) [inline spyderlib/restart_app.py]: Same here for the `for` loop and the `else`

File: spyderlib/restart_app.py [hunk 12]
```diff
@@ -129,18 +180,60 @@ def main():
...
+    # -------------------------------------------------------------------------
+    if reset_ok:
+        # Try to restart
+        try:
+            p = subprocess.Popen(command, shell=shell, env=env)
+        except Exception as error:
+            p.kill()
+            error_message(type_='restart', error=error)
+    else:
+        error_message(type_='reset')
```
Comment (Nodd) [inline spyderlib/restart_app.py]: Why do this here and not when you set `reset_ok = False` ?

File: spyderlib/restart_app.py [hunk 13]
```diff
@@ -129,18 +181,51 @@ def main():
     shell = os.name != 'nt'
 
     # Wait for original process to end before launching the new instance
-    while True:
+    for counter in range(range(60*5/SLEEP_TIME)):  # Number in seconds (60*5)
```
Comment (Nodd) [inline spyderlib/restart_app.py]: Too many ranges here !

File: spyderlib/restart_app.py (hunk 11 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: I have fixed this already

File: spyderlib/restart_app.py (hunk 12 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Cause I have two places to define `reset_ok = False`, one for an exception and the other for timeout... so I guess this looked cleaner to me as opposed, to writing `error_message(type_='reset')` in both places (where I define `reset_ok = False`)

File: spyderlib/restart_app.py (hunk 13 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Shit!

File: spyderlib/restart_app.py [hunk 14]
```diff
@@ -129,18 +181,46 @@ def main():
...
+            p = subprocess.Popen(command_reset, shell=shell)
+            p.communicate()
+            pid_reset = p.pid
+        except Exception as error:
+            p.kill()
+            launch_error_message(type_=RESET_ERROR, error=error)
+
+        # Wait for reset process to end before restarting
+        for counter in range(60/SLEEP_TIME):  # Number in seconds (60)
+            if not is_pid_running(pid_reset):
```
Comment (Nodd) [inline spyderlib/restart_app.py]: If there was an exception before, `pid_reset` won't be defined.
I think this block should go in an `else:` statement after the `try:` and `except:`, since you shouldn''t have to wait if there was an error (I guess).

File: spyderlib/restart_app.py [hunk 15]
```diff
@@ -129,18 +181,46 @@ def main():
...
+    # -------------------------------------------------------------------------
+    if reset:
+        command_reset = '"{0}" "{1}" {2}'.format(python, spyder, args_reset)
+
+        try:
+            p = subprocess.Popen(command_reset, shell=shell)
+            p.communicate()
+            pid_reset = p.pid
+        except Exception as error:
+            p.kill()
```
Comment (Nodd) [inline spyderlib/restart_app.py]: `p` could be undefined if the error comes from `Popen`.
Error checking is complicated, even with exceptions !

**Comment by Nodd** [comment]
Sorry for all the nitpicks...

File: spyderlib/restart_app.py (hunk 14 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Indeed, I will make the adjustments, good suggestions

**Comment by goanpeca** [comment]
@Nodd, do not worry about the nitpicks, I prefer comments to no comment at all !

**Comment by goanpeca** [comment]
@Nodd no more nitpicking :-) ?

**Comment by Nodd** [comment]
Nope, It looks fine ! :)
I didn't test it for real, however.

**Comment by goanpeca** [comment]
@ccordoba12 any comments here?

File: spyderlib/restart_app.py [hunk 16]
```diff
@@ -20,6 +20,8 @@
 
 
 PY2 = sys.version[0] == '2'
+CLOSE_ERROR, RESET_ERROR, RESTART_ERROR = list(range(1, 4, 1))
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: Please add a comment above this line to explan why this number error codes mean

File: spyderlib/restart_app.py [hunk 17]
```diff
@@ -63,6 +65,47 @@ def is_pid_running(pid):
         return _is_pid_running_on_unix(pid)
 
 
+def launch_error_message(type_, error=None):
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: `type_` seems an ugly error variable name to me. Perhaps `error_type` would be better?

File: spyderlib/restart_app.py [hunk 18]
```diff
@@ -63,6 +65,47 @@ def is_pid_running(pid):
...
+    ----------
+    type_ : int [CLOSE_ERROR, RESET_ERROR, RESTART_ERROR]
+    """
+    from spyderlib.baseconfig import _
+    from spyderlib.qt.QtGui import QMessageBox, QDialog
+    from spyderlib.utils import icon_manager as ima
+    from spyderlib.utils.qthelpers import qapplication
+
+    messages = {CLOSE_ERROR: _("The previous instance of Spyder has not "
+                               "closed.\nRestart aborted."),
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: Perhaps a better message for this error would be:

    It was not possible to close the previous Spyder instance.
    Restart aborted

File: spyderlib/restart_app.py [hunk 19]
```diff
@@ -63,6 +65,47 @@ def is_pid_running(pid):
...
+                RESET_ERROR: _("Spyder could not reset to factory defaults.\n"
+                               "Restart aborted."),
+                RESTART_ERROR: _("Spyder could not restart.\n"
+                                 "Restart aborted.")}
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: Another suggestion

    It was not possible to restart Spyder.
    Operation aborted.

File: spyderlib/restart_app.py [hunk 20]
```diff
@@ -129,18 +181,47 @@ def main():
     shell = os.name != 'nt'
 
     # Wait for original process to end before launching the new instance
-    while True:
+    wait_time = 60*5  # Seconds
+    for counter in range(int(wait_time/SLEEP_TIME)):
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: What does this `for` loop exactly do? Could you add a comment below

    # Wait for original process ...

to say something about it?

**Comment by ccordoba12** [comment]
Ok, first review done.

**Comment by ccordoba12** [comment]
@goanpeca, about your first TODO: I think both locations are fine.

File: spyderlib/restart_app.py (hunk 20 above)
Comment (Nodd) [inline spyderlib/restart_app.py]: There is a comment one line above :)

Comment (ccordoba12) [inline spyderlib/restart_app.py]: Yeah, but I'd like to see a more detailed explanation.

**Comment by ccordoba12** [comment]
@goanpeca, could you address my comments so we can merge this one?

**Comment by goanpeca** [comment]
Oops.. I did something stupid

**Comment by goanpeca** [comment]
@ccordoba12 fixed!

File: spyderlib/restart_app.py [hunk 21]
```diff
@@ -20,6 +20,11 @@
 
 
 PY2 = sys.version[0] == '2'
+SLEEP_TIME = 0.2  # Seconds for throttling control
+
+# Possible error codes when restarting/reseting spyder
+# Start in 1 cause the error code is used in an if clause
+CLOSE_ERROR, RESET_ERROR, RESTART_ERROR = list(range(1, 4, 1))
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: This is bit weird to read :-) What about

```python
CLOSE_ERROR, RESET_ERROR, RESTART_ERROR = [1, 2, 3]
```

File: spyderlib/restart_app.py [hunk 22]
```diff
@@ -63,6 +68,50 @@ def is_pid_running(pid):
...
+    from spyderlib.baseconfig import _
+    from spyderlib.qt.QtGui import QMessageBox, QDialog
+    from spyderlib.utils import icon_manager as ima
+    from spyderlib.utils.qthelpers import qapplication
+
+    messages = {CLOSE_ERROR: _("It was not possible to close the previous "
+                               "Spyder instance.\nRestart aborted."),
+                RESET_ERROR: _("Spyder could not reset to factory defaults.\n"
+                               "Restart aborted."),
+                RESTART_ERROR: _("It was not possible restart Spyder.\n"
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: It was not possible **to** restart

File: spyderlib/restart_app.py [hunk 23]
```diff
@@ -89,29 +138,38 @@ def main():
...
         error = "This script can only be called from within a Spyder instance"
         raise RuntimeError(error)
 
     # Variables were stored as string literals in the environment, so to use
     # them we need to parse them in a safe manner.
     is_bootstrap = ast.literal_eval(is_bootstrap)
-    pid = int(pid)
+    pid = ast.literal_eval(pid)
     args = ast.literal_eval(spyder_args)
+    reset = ast.literal_eval(reset)
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: What `reset` evals to? `True` or `False`? Just curious :-)

File: spyderlib/restart_app.py [hunk 24]
```diff
@@ -128,19 +186,52 @@ def main():
...
+    # time and check, otherwise an error is launched
+    wait_time = 60*5  # Seconds
+    for counter in range(int(wait_time/SLEEP_TIME)):
         if not is_pid_running(pid):
             break
-        time.sleep(0.2)  # Throttling control
+        time.sleep(SLEEP_TIME)  # Throttling control
+    else:
+        # The old spyder instance took too long to close and restart aborts
+        launch_error_message(type_=CLOSE_ERROR)
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: Now the arg is called `error_type`, not `type_`

File: spyderlib/restart_app.py [hunk 25]
```diff
@@ -128,19 +186,52 @@ def main():
     # Adjust the command and/or arguments to subprocess depending on the OS
     shell = os.name != 'nt'
 
-    # Wait for original process to end before launching the new instance
-    while True:
+    # Before launching a new Spyder instance we need to make sure that the
+    # previous one has closed. We wait for a fixed and "reasonable" amount of
+    # time and check, otherwise an error is launched
+    wait_time = 60*5  # Seconds
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: So we're waiting for 5 min. Isn't that way too much time? I mean, specially if we are not giving our users any visual feedback about what's happening :-)

Two questions:
1. Isn't 2 or 3 min at most enough?
2. Can't we show a throbber or something while doing this? I mean, something like we already have when creating kernels for our IPython consoles?

File: spyderlib/restart_app.py (hunk 23 above)
Comment (ccordoba12) [inline spyderlib/restart_app.py]: No problem, I just saw below that this is the case ;-)

File: spyderlib/restart_app.py (hunk 24 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Oops

File: spyderlib/restart_app.py (hunk 21 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Fair enough

File: spyderlib/restart_app.py (hunk 22 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: Oops! 2 am coding!

File: spyderlib/restart_app.py (hunk 23 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: :-)

File: spyderlib/restart_app.py (hunk 25 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: 1. I guess, but windows sometimes takes sooooooo long!
2. Sure we could, what would you like to see? a "splash screen saying, Spyder is restarting and the throbber?

**Comment by goanpeca** [comment]
@ccordoba12 fixed your comments and updated the code to show the spyder Splash screen with animated ellipsis.

**Comment by ccordoba12** [comment]
This is quite neat!! Thanks a lot for following my suggestion ;-)

One last suggestion: Would it be possible to offer our users the possibility to drop the resetting and restarting process, or continue waiting? This could be done with a `QMessageBox` after 1 min on Linux/Mac and 2 min on Windows.

What do you think?

**Comment by goanpeca** [comment]
Sure we can do that, do you want this as a separate PR? or into this one?

**Comment by ccordoba12** [comment]
I'd like to have it in this one, please :-)

File: spyderlib/restart_app.py [hunk 26]
```diff
@@ -84,34 +93,116 @@ def to_text_string(obj, encoding=None):
...
     spyder_args = os.environ.pop('SPYDER_ARGS', None)
     pid = os.environ.pop('SPYDER_PID', None)
     is_bootstrap = os.environ.pop('SPYDER_IS_BOOTSTRAP', None)
+    reset = os.environ.pop('SPYDER_RESET', None)
 
     # Get the spyder base folder based on this file
     spyder_folder = osp.split(osp.dirname(osp.abspath(__file__)))[0]
 
-    if any([not spyder_args, not pid, not is_bootstrap]):
+    if any([not spyder_args, not pid, not is_bootstrap, not reset]):
```
Comment (SylvainCorlay) [inline spyderlib/restart_app.py]: (not a) or (not b) or (not c) == not (a and b and c)

File: spyderlib/restart_app.py [hunk 27]
```diff
@@ -128,19 +219,72 @@ def main():
...
     env = os.environ.copy()
+
+    # Reset Spyder (if required)
+    # -------------------------------------------------------------------------
+    if reset:
+        restarter.set_splash_message(_('Resetting Spyder to defaults'))
+        command_reset = '"{0}" "{1}" {2}'.format(python, spyder, args_reset)
+
+        try:
+            p = subprocess.Popen(command_reset, shell=shell)
```
Comment (SylvainCorlay) [inline spyderlib/restart_app.py]: env=env

File: spyderlib/restart_app.py (hunk 26 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: True :)

File: spyderlib/restart_app.py (hunk 27 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: That fixed it ?

Comment (SylvainCorlay) [inline spyderlib/restart_app.py]: neveermind.

Comment (goanpeca) [inline spyderlib/restart_app.py]: ehhh... what ?

File: spyderlib/spyder.py [hunk 28]
```diff
@@ -2735,6 +2754,7 @@ def restart(self):
         env['SPYDER_ARGS'] = spyder_args
         env['SPYDER_PID'] = str(pid)
         env['SPYDER_IS_BOOTSTRAP'] = str(is_bootstrap)
+        env['SPYDER_RESET'] = str(reset)
```
Comment (SylvainCorlay) [inline spyderlib/spyder.py]: adding `env['PYTHON_PATH'] = sys.path` gets rid of the import errors, but it seems the old Spyder is not shutting down though.

Comment (goanpeca) [inline spyderlib/spyder.py]: So the problem was only with the reset then,, or even if you just try to restart spyder it was failing?

Comment (SylvainCorlay) [inline spyderlib/spyder.py]: @blink1073 can you give it a shot in windows?

File: spyderlib/restart_app.py [hunk 29]
```diff
@@ -8,8 +8,9 @@
 """
 Restart Spyder
 
-A helper script that allows Spyder to restart from within the application.
-"""
+A helper script that allows to restart (and also reset) Spyder from within the
+running application.
+my """
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: It seems to be an error here.

File: spyderlib/spyder.py (hunk 28 above)
Comment (goanpeca) [inline spyderlib/spyder.py]: So it was working on windows? @blink1073 ?

Comment (blink1073) [inline spyderlib/spyder.py]: It works with the latest version on Windows.

Comment (goanpeca) [inline spyderlib/spyder.py]: Great :-), thanks for checking.

File: spyderlib/restart_app.py [hunk 30]
```diff
@@ -126,21 +232,63 @@ def main():
...
+        except Exception as error:
+            restarter.launch_error_message(error_type=RESET_ERROR, error=error)
+        else:
+            p.communicate()
+            pid_reset = p.pid
+
+        # Before launching a new Spyder instance we need to make sure that the
+        # reset subprocess has closed. We wait for a fixed and "reasonable"
+        # amount of time and check, otherwise an error is launched.
+        wait_time = 60  # Seconds
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: So we're waiting the same amount of time in all OSes? I thought we decided to wait more on Windows (90 sec) and less on Linux and Mac (30 sec).

File: spyderlib/spyder.py [hunk 31]
```diff
@@ -2735,6 +2754,11 @@ def restart(self):
         env['SPYDER_ARGS'] = spyder_args
         env['SPYDER_PID'] = str(pid)
         env['SPYDER_IS_BOOTSTRAP'] = str(is_bootstrap)
+        env['SPYDER_RESET'] = str(reset)
+        if os.name == 'nt':
+            env['PYTHONPATH'] = ';'.join(sys.path)
+        else:
+            env['PYTHONPATH'] = ':'.join(sys.path)
```
Comment (ccordoba12) [inline spyderlib/spyder.py]: Do we need to the full `sys.path` here? Or just its first entry?

Comment (goanpeca) [inline spyderlib/spyder.py]: Hmmm I guess the first entry should be enough, is it a problem that we use all anyway?

File: spyderlib/restart_app.py (hunk 30 above)
Comment (goanpeca) [inline spyderlib/restart_app.py]: This is the reset process, the reset process normally is fast on windows as well. Is on waiting or the spyder instance to close that takes long, [check this](https://github.com/spyder-ide/spyder/pull/2423/files#diff-6614a100b3eb58a6c02b7b1da9e72fa1R240)

Comment (blink1073) [inline spyderlib/restart_app.py]: I'd say we don't know exactly what order they added things to the path.  Worst case we have duplicated entries with the current method.

Comment (ccordoba12) [inline spyderlib/restart_app.py]: So what you say is that we should leave `sys.path` as it is?

Comment (blink1073) [inline spyderlib/restart_app.py]: I'm saying leave the PR as it is.  :smile:

Comment (goanpeca) [inline spyderlib/restart_app.py]: I concur with blinky

File: spyderlib/spyder.py (hunk 31 above)
Comment (ccordoba12) [inline spyderlib/spyder.py]: This could cause issues with our consoles. Specially with our Mac app because it does some funky things to PYTHONPATH.

Question: I think we need this only for `bootstrap` to work, right @goanpeca?

Comment (goanpeca) [inline spyderlib/spyder.py]: Yep, I would say that is correct. If not running from bootstrap then it must be running from an installed version (site-packages) so the imports should be fine then.

I can make add an if so that the path is updated only if running from bootstrap?

Comment (ccordoba12) [inline spyderlib/spyder.py]: Even easier: we could `DEV` here like this:

```python
if DEV
    if os.name == 'nt':
        env['PYTHONPATH'] = ';'.join(sys.path)
    else:
        env['PYTHONPATH'] = ':'.join(sys.path)
```

Comment (goanpeca) [inline spyderlib/spyder.py]: Yes, that is what I meant...

Comment (ccordoba12) [inline spyderlib/spyder.py]: Ok, nice!

File: spyderlib/restart_app.py [hunk 32]
```diff
@@ -126,21 +232,63 @@ def main():
...
     # Adjust the command and/or arguments to subprocess depending on the OS
-    shell = os.name != 'nt'
+    shell = not IS_WINDOWS
 
-    # Wait for original process to end before launching the new instance
-    while True:
+    # Before launching a new Spyder instance we need to make sure that the
+    # previous one has closed. We wait for a fixed and "reasonable" amount of
+    # time and check, otherwise an error is launched
+    wait_time = 60*5 if IS_WINDOWS else 60*3 # Seconds
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: So we discussed this with @blink1073 and @SylvainCorlay and all of us agree that these waiting times are way too big. We think nobody is going to wait 3 min, much less 5 min, for Spyder to start :-)

So our proposal for this is: 30 sec for Linux/Mac (perhaps even 20), and 75 sec (90 tops) for Windows.

Comment (goanpeca) [inline spyderlib/restart_app.py]: Ok Done...

30 and 90

**Comment by goanpeca** [comment]
Any my mental health?

**Comment by goanpeca** [comment]
Anything holding this @ccordoba12?

File: spyderlib/restart_app.py [hunk 33]
```diff
@@ -84,34 +96,128 @@ def to_text_string(obj, encoding=None):
...
+                    RESET_ERROR: _("Spyder could not reset to factory "
+                                   "defaults.\nRestart aborted."),
+                    RESTART_ERROR: _("It was not possible to restart Spyder.\n"
+                                     "Operation aborted.")}
+        titles = {CLOSE_ERROR: _("Spyder exit error"),
+                  RESET_ERROR: _("Spyder reset error"),
+                  RESTART_ERROR: _("Spyder restart error")}
+
+        if error:
+            e = error.__repr__()
```
Comment (ccordoba12) [inline spyderlib/restart_app.py]: Why is this necessary?

**Comment by ccordoba12** [comment]
I think waiting 60 secs for `--reset` to work is too much because that would be added to the restart time, right?

Also, I'd like to discuss how the logic for reset + restart works. So `restart_app` calls `spyder --reset` first and then it closes the app, or is the other way around?

**Comment by goanpeca** [comment]
Ok. When you restart first we wait for spyder to close. If that happenned before the max wait time then we launch the reset process, if that went fine and took less than the stated time then the new spyd instance is launched.

**Comment by ccordoba12** [comment]
So this means a user could have to wait up to (30 or 90) + 60 secs for a reset (plus the time it take for a new instance to appear).

Can we reduce the time reset to 20 secs or so? It usually doesn't take that long ;-)

**Comment by goanpeca** [comment]
Yes, you are correct.

Sure, 20 secs is fine (I guess... we will know otherwise!)

**Comment by goanpeca** [comment]
Anything besides that ?

**Comment by goanpeca** [comment]
Anything else @ccordoba12 ?

**Comment by ccordoba12** [comment]
Ok, merging this with @goanpeca (in person) for his own mental health :-P

**Comment by blink1073** [comment]
woot!
